README.rst
=======
Hirola
=======

.. image::
    https://img.shields.io/pypi/pyversions/cslug?label=Python&color=%23184159
    :alt: PyPI version
    :target: https://pypi.org/project/hirola/

.. image:: https://img.shields.io/badge/coverage-100%25-%23184159

NumPy vectorized hash table written in C for fast (roughly 10x faster) ``set``/``dict``
like operations.

* Free software: `MIT license <https://github.com/bwoodsend/hirola/blob/master/LICENSE>`_
* Documentation: `<https://hirola.readthedocs.io/>`_
* Source code: `<https://github.com/bwoodsend/hirola/>`_
* Releases: `<https://pypi.org/project/hirola/>`_

A ``hirola.HashTable`` is to ``dict`` what ``numpy.array`` is to ``list``.
By imposing some constraints, vectorising, and translating into C, the speed
can be improved dramatically.
For hirola, these constraints are:

* Keys must all be of the same predetermined type and size.
* The maximum size of a table must be chosen in advance and managed explicitly.
* To get any performance boost, operations should be done in bulk.
* Elements can not be removed.

If any of the above are not satisfied for your use case then don't use
hirola.


Installation
------------

Install Hirola with pip:

.. code-block:: console

    pip install hirola


Quickstart
----------

``HashTable``
*************

At the rawest and dirtiest level lives the ``HashTable`` class.
A ``HashTable`` can be though of as a ``dict`` but with only an enumeration for
values.
To construct an empty hash table:

.. code-block:: python

    import numpy as np
    from hirola import HashTable

    table = HashTable(
        20,  # <--- Maximum size for the table - up to 20 keys.
        "U10",  # <--- NumPy dtype - strings of up to 10 characters.
    )

Keys may be added individually...

.. code-block:: python

    >>> table.add("cat")
    0

... But it's much more efficient to add in bulk.
The return value is an enumeration of when each key was first added.
Duplicate keys are not re-added.

.. code-block:: python

    >>> table.add(["dog", "cat", "moose", "gruffalo"])
    array([1, 0, 2, 3])


Multidimensional inputs give multidimensional outputs of matching shapes.

.. code-block:: python

    >>> table.add([["rabbit", "cat"],
    ...            ["gruffalo", "moose"],
    ...            ["werewolf", "gremlin"]])
    array([[4, 0],
           [3, 2],
           [5, 6]])

Inspect all keys added so far via the ``keys`` attribute.
(Note that, unlike ``dict.keys()``, it's a property instead of a method.)

.. code-block:: python

    >>> table.keys
    array(['cat', 'dog', 'moose', 'gruffalo', 'rabbit', 'werewolf', 'gremlin'],
          dtype='<U10')

Key indices can be retrieved with ``table.get(key)`` or just ``table[key]``.
Again, retrieval is NumPy vectorised and is much faster if given large arrays of
inputs rather than one at a time.

.. code-block:: python

    >>> table.get("dog")
    1
    >>> table[["moose", "gruffalo"]]
    array([2, 3])

Like the Python dict,
using ``table[key]`` raises a ``KeyError`` if keys are missing
but using ``table.get(key)`` returns a configurable default.
Unlike Python's dict, the default is ``-1``.

.. code-block:: python

    >>> table["tortoise"]
    KeyError: "key = 'tortoise' is not in this table."
    >>> table.get("tortoise")
    -1
    >>> table.get("tortoise", default=99)
    99
    >>> table.get(["cat", "bear", "tortoise"], default=[100, 101, 102])
    array([  0, 101, 102])


Choosing a ``max`` size
.......................

Unlike Python's ``set`` and ``dict``, ``Hirola`` does not manage its size
automatically.
To prevent wasted resizing (which is what Python does under the hood),
you have full control of and responsibility for how much space the table uses.
Obviously the table has to be large enough to fit all the keys in it.
Additionally, when a hash table gets to close to full it becomes much slower.
Depending on how much you favour speed over memory you should add 20-50% extra
headroom.
If you intend to a lot of looking up of the same small set of values then it can
continue to run faster if you increase ``max`` to 2-3x its minimal size.


Structured key data types
.........................

To indicate that an array axis should be considered as a single key,
use NumPy's structured dtypes.
In the following example, the data type ``(points.dtype, 3)``
indicates that a 3D point - a triplet of floats -
should be considered as one object.
See ``help(HashTable.dtype)`` for more information of specifying dtypes.
Only the last axis or last axes may be thought of as single keys.
For other setups, first convert with ``numpy.transpose()``.

.. code-block:: python

    import numpy as np
    from hirola import HashTable

    # Create a cloud of 3D points with duplicates. This is 3000 points in total,
    # with up to 1000 unique points.
    points = np.random.uniform(-30, 30, (1000, 3))[np.random.choice(1000, 3000)]

    # Create an empty hash table.
    # In practice, you generally don't know how many unique elements there are
    # so we'll pretend we don't either an assume the worst case of all 3000 are
    # unique. We'll also give 25% padding for speed.
    table = HashTable(len(points) * 1.25, (points.dtype, 3))

    # Add all points to the table.
    ids = table.add(points)

Duplicate-free contents can be accessed from ``table.keys``:

.. code-block:: python

    >>> table.keys  # <--- These are `points` but with no duplicates.
    array([[  3.47736554, -15.17112511,  -9.51454466],
           [ -6.46948046,  23.64504329, -16.25743105],
           [-27.02527253, -16.1967225 , -10.11544157],
           ...,
           [  3.75972597,   1.24130412,  -8.14337206],
           [-13.62256791,  11.76551455, -13.31312988],
           [  0.19851678,   4.06221179, -22.69006592]])
    >>> table.keys.shape
    (954, 3)

Each point's location in ``table.keys`` is returned by ``table.add()``,
similarly to ``numpy.unique(..., return_args=True)``.

.. code-block:: python

    >>> ids  # <--- These are the indices in `table.keys` of each point in `points`.
    array([  0,   1,   2, ..., 290, 242, 669])
    >>> np.array_equal(table.keys[ids], points)
    True

Lookup the indices of points without adding them using ``table.get()``.


Recipes
*******

A ``HashTable`` can be used to replicate a `dict <as-a-dict>`_,
`set <as-a-set>`_ or a `collections.Counter <as-a-collections.Counter>`_.
These might turn into their own proper classes in the future or they might not.


.. _as-a-dict:

Using a ``HashTable`` as a ``dict``
...................................

A ``dict`` requires a second array for values.
The output of ``HashTable.add()``  and ``HashTable.get()`` should be used as
indices of ``values``:

.. code-block:: python

    import numpy as np
    from hirola import HashTable

    # The `keys` - will be populated with names of African countries.
    countries = HashTable(40, (str, 20))
    # The `values` - will be populated with the names of each country's capital city.
    capitals = np.empty(countries.max, (str, 20))

Add or set items using the pattern ``values[table.add(key)] = value``:

.. code-block:: python

    capitals[countries.add("Algeria")] = "Al Jaza'ir"

Or in bulk:

.. code-block:: python

    new_keys = ["Angola", "Botswana", "Burkina Faso"]
    new_values = ["Luanda", "Gaborone", "Ouagadougou"]
    capitals[countries.add(new_keys)] = new_values

Like Python dicts, overwriting values is exactly the same as writing them.

Retrieve values with ``values[table[key]]``:

.. code-block:: python

    >>> capitals[countries["Botswana"]]
    'Gaborone'
    >>> capitals[countries["Botswana", "Algeria"]]
    array(['Gaborone', "Al Jaza'ir"], dtype='<U20')

View all keys and values with ``table.keys`` and ``values[:len(table)]``.
A ``HashTable`` remembers the order keys were first added so this dict is
automatically a sorted dict.

.. code-block:: python

    # keys
    >>> countries.keys
    array(['Algeria', 'Angola', 'Botswana', 'Burkina Faso'], dtype='<U20')
    # values
    >>> capitals[:len(countries)]
    array(["Al Jaza'ir", 'Luanda', 'Gaborone', 'Ouagadougou'], dtype='<U20')

Depending on the usage scenario,
it may or may not make sense to want an equivalent to  ``dict.items()``.
If you do want an equivalent,
use ``numpy.rec.fromarrays([table.keys, values[:len(table)]])``,
possibly adding a ``names=`` option:

.. code-block:: python

    >>> np.rec.fromarrays([countries.keys, capitals[:len(countries)]],
    ...                   names="countries,capitals")
    rec.array([('Algeria', "Al Jaza'ir"), ('Angola', 'Luanda'),
               ('Botswana', 'Gaborone'), ('Burkina Faso', 'Ouagadougou')],
              dtype=[('countries', '<U20'), ('capitals', '<U20')])

If the keys and values have the same dtype then ``numpy.c_`` works too.

.. code-block:: python

    >>> np.c_[countries.keys, capitals[:len(countries)]]
    array([['Algeria', "Al Jaza'ir"],
           ['Angola', 'Luanda'],
           ['Botswana', 'Gaborone'],
           ['Burkina Faso', 'Ouagadougou']], dtype='<U20')


.. _as-a-set:

Using a ``HashTable`` as a ``set``
..................................

To get set-like capabilities from a ``HashTable``,
leverage the ``contains()`` method.
For these examples we will experiment with integer multiples of 3 and 7.

.. code-block:: python

    import numpy as np

    of_3s = np.arange(0, 100, 3)
    of_7s = np.arange(0, 100, 7)

We'll only require one array to be converted into a hash table.
The other can remain as an array.
If both are hash tables, simply use one table's ``keys`` attribute as the array.

.. code-block:: python

    from hirola import HashTable

    table_of_3s = HashTable(len(of_3s) * 1.25, of_3s.dtype)
    table_of_3s.add(of_3s)

Use ``table.contains()`` as a vectorised version of ``in``.

.. code-block:: python

    >>> table_of_3s.contains(of_7s)
    array([ True, False, False,  True, False, False,  True, False, False,
            True, False, False,  True, False, False])

From the above, the common set operations can be derived with following:

*   ``set.intersection()`` - Values in the array and in the set:

.. code-block:: python

        >>> of_7s[table_of_3s.contains(of_7s)]
        array([ 0, 21, 42, 63, 84])

*   Set subtraction - Values in the array which are not in the set:

.. code-block:: python

        >>> of_7s[~table_of_3s.contains(of_7s)]
        array([ 7, 14, 28, 35, 49, 56, 70, 77, 91, 98])

*   ``set.union()`` - Values in either the table or in the tested array (with no
    duplicates):

.. code-block:: python

        >>> np.concatenate([table_of_3s.keys, of_7s[~table_of_3s.contains(of_7s)]], axis=0)
        array([ 0,  3,  6,  9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39, 42, 45, 48,
               51, 54, 57, 60, 63, 66, 69, 72, 75, 78, 81, 84, 87, 90, 93, 96, 99,
                7, 14, 28, 35, 49, 56, 70, 77, 91, 98])


.. _`as-a-collections.Counter`:

Using a ``HashTable`` as a ``collections.Counter``
..................................................

For this example,
let's give ourselves something a bit more substantial to work on.
Counting word frequencies in Shakespeare's Hamlet play is the
trendy example for ``collections.Counter`` and it's what we'll use too.

.. code-block:: python

    from urllib.request import urlopen
    import re
    import numpy as np

    hamlet = urlopen("https://gist.githubusercontent.com/provpup/2fc41686eab7400b796b/raw/b575bd01a58494dfddc1d6429ef0167e709abf9b/hamlet.txt").read()
    words = np.array(re.findall(rb"([\w']+)", hamlet))

A counter is just a ``dict`` with integer values and a ``dict`` is just a hash
table with a separate array for values.

.. code-block:: python

    from hirola import HashTable

    word_table = HashTable(len(words), words.dtype)
    counts = np.zeros(word_table.max, dtype=int)

The only new functionality that is not defined in `using a hash table as a dict
<as-a-dict>`_ is the ability to count keys as they are added.
To count new elements use the rather odd line
``np.add(counts, table.add(keys), 1)``.

.. code-block:: python

    np.add.at(counts, word_table.add(words), 1)

This line does what you might expect ``counts[word_table.add(words)] += 1`` to
do but, due to the way NumPy works,
the latter form fails to increment each count more than once if ``words``
contains duplicates.

Use NumPy's indirect sorting functions to get most or least common keys.

.. code-block:: python

    # Get the most common word.
    >>> word_table.keys[counts[:len(word_table)].argmax()]
    b'the'

    # Get the top 10 most common words. Note that these are unsorted.
    >>> word_table.keys[counts[:len(word_table)].argpartition(-10)[-10:]]
    array([b'it', b'and', b'my', b'of', b'in', b'a', b'to', b'the', b'I',
           b'you'], dtype='|S14')

    # Get all words in ascending order of commonness.
    >>> word_table.keys[counts[:len(word_table)].argsort()]
    array([b'END', b'whereat', b"griev'd", ..., b'to', b'and', b'the'],
          dtype='|S14')



A Security Note
---------------

Unlike the builtin ``hash()`` used internally by Python's ``set`` and ``dict``,
``hirola`` does not randomise a hash seed on startup
making an online server running ``hirola`` more vulnerable to denial of service
attacks.
In such an attack, the attacker clogs up your server by sending it requests that
he/she knows will cause hash collisions and therefore slow it down.
Whereas a Python hash table's size is always predictably the next power of 8
above ``len(table) * 3 / 2``, a ``hirola.HashTable()`` may be any size meaning
that you can make an attack considerably more difficult by adding a little
randomness to the sizes of your hash tables.
But if your writing an online server
which performs dictionary lookup based on user input
and your user-base doesn't like you much
or you have some very spiteful below-the-belt competitors
then I recommend that you don't use this library.


Credits
-------

This package was initially created with Cookiecutter_ and a fork of the
`audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=========
Changelog
=========

Release history for :mod:`hirola`.

v0.3.0
------

*   Add :ref:`handling of nearly full hash tables`.

*   Add an **in_place** option to `HashTable.resize`.

*   Improve performance by a few percent.


v0.2.1
------

*   Fix macOS wheels to include ARM64 binaries as well a x86_64.


v0.2.0
-------

*   Include precompiled 32 bit wheels for Windows.


v0.1.0
------

Initial release.


hirola/_hash_table.py
# -*- coding: utf-8 -*-
"""
"""

import numbers
import ctypes
import math

from numbers import Number
from typing import Union, Tuple

import numpy as np
from cslug import CSlug, ptr, anchor, Header

hashes_header = Header(*anchor("hashes.h", "hashes.c"),
                       includes=["<stddef.h>", "<stdint.h>"])
slug = CSlug(anchor("hash_table", "hash_table.h", "hash_table.c", "hashes.c"),
             headers=hashes_header)

dtype_types = Union[np.dtype, np.generic, type, str, list, tuple]


class HashTable(object):
    """The raw core behind a set or a dictionary's keys.

    A hash table resembles a dictionary where its keys are the `keys`
    array but the values are an just an enumeration. It's core API:

    * Use `add` to add new keys if they've not been already added.
    * The `keys` lists all keys that have been added in the order that
      they were added.
    * Use `get` to retrieve indices of keys in `keys`.

    """
    _keys: np.ndarray
    _NO_DEFAULT = object()

    def __init__(self, max: Number, dtype: dtype_types,
                 almost_full=(.9, "warn")):
        """

        Args:
            max:
                An upper bound for the number of keys which can fit in this
                table. Sets the `max` attribute.
            dtype:
                The data type for the table's keys. Sets the `dtype`
                attribute.
            almost_full:
                The handling of almost full hash tables. Sets the `almost_full`
                attribute.

        The **max** parameter is silently normalised to `int` and clipped
        to a minimum of 1 if it is less than 1.

        """
        self._dtype = np.dtype(dtype)
        key_size = self.dtype.itemsize
        self._base_dtype, self._dtype_shape = self.dtype.base, self.dtype.shape
        if self._base_dtype == object:
            raise TypeError("Object arrays are not permitted.")
        if self._base_dtype.kind in "SUV":
            # String-like types are checked differently.
            self._check_dtype = self._check_str_dtype

        if max <= 0:
            # Zero-sized tables get in the way of modulo.
            # Negative-sized tables obviously don't make sense.
            max = 1
        max = int(max)

        self._hash_owners = np.full(max, -1, np.intp)
        self._keys = np.empty(max, dtype=self.dtype)
        self._keys_readonly = np.frombuffer(self._keys, self.dtype)
        self._keys_readonly.flags.writeable = False

        hash = choose_hash(key_size)
        self._destroyed = False
        self._raw = slug.dll.HashTable(max, key_size, ptr(self._hash_owners),
                                       ptr(self._keys_readonly),
                                       hash=ctypes.cast(hash, ctypes.c_void_p))
        self.almost_full = almost_full

    @property
    def max(self) -> int:
        """The maximum number of elements allowed in this table.

        Adding keys to exceed this maximum will trigger a
        :class:`~hirola.exceptions.HashTableFullError`. Nearly full tables
        will `add` and `get` much slower. For best performance,
        choose a maximum size which is 25-50% larger than you expect it to get.

        """
        return self._raw.max

    @property
    def dtype(self) -> np.dtype:
        """The data type for the table's keys.

        Use a structured `numpy.dtype` to indicate that several numbers
        make up a single key.

        Examples:

            - Individual float keys: :py:`HashTable(100, np.float64)`.
            - Triplets of floats keys: :py:`HashTable(100, (np.float64, 3))`.
            - Strings of up to 20 characters: :py:`HashTable(100, (str, 20))`.
            - Mixed types records data:
              :py:`HashTable(100, [("firstname", str, 20), ("lastname", str, 20), ("age", int)])`.

        """
        return self._dtype

    @property
    def key_size(self) -> int:
        """The number of bytes per key."""
        return self._raw.key_size

    @property
    def keys(self) -> np.ndarray:
        """The unique elements in the table.

        Unlike Python's builtin `dict`, this `keys` is a
        `property` rather than a method. Keys must be immutable hence
        this array is a readonly view. See `destroy` to release the
        writable version.

        """
        return self._keys_readonly[:self.length]

    @property
    def length(self) -> int:
        """The number of elements currently in this table. Aliased via
        :py:`len(table)`."""
        return self._raw.length

    def __len__(self):
        return self.length

    def _add(self, key):
        return slug.dll.HT_add(self._raw._ptr, ptr(key))

    def _get(self, key):
        return slug.dll.HT_get(self._raw._ptr, ptr(key))

    def add(self, keys) -> np.ndarray:
        """Add **keys** to the table.

        Any key which is already in `keys` is not added again. Returns
        the index of each key in `keys` similarly to `get`.

        Raises:
            ValueError:
                If the :attr:`~numpy.ndarray.dtype` of **keys** doesn't match
                the `dtype` of this table.
            exceptions.HashTableFullError:
                If there is no space to place new keys.
            exceptions.HashTableDestroyed:
                If the `destroy` method has been previously called.
            exceptions.AlmostFull:
                If the table becomes nearly full and is configured to raise an
                error (set by the `almost_full` attribute).
        Warns:
            exceptions.AlmostFull:
                If the table becomes nearly full and is configured to warn (set
                by the `almost_full` attribute).

        """
        self._check_destroyed()

        keys, shape = self._norm_input_keys(keys)
        out = np.empty(shape, np.intp)

        # This while loop will only iterate a second time if the "almost full"
        # threshold is enabled and crossed. It will only iterate more than twice
        # if `self.almost_full` is set to automatically upsize the table.
        index = -1
        while True:
            index = slug.dll.HT_adds(self._raw._ptr, ptr(keys), ptr(out),
                                     out.size, index + 1)

            # If everything worked. Return the indices.
            if index == out.size:
                return out if shape else out.item()

            # If the `almost_full` threshold has been crossed:
            if index < 0:
                # Convert to a real positive index.
                index = -1 - index
                # Issue a warning or raise an error or resize the table as per
                # the user's configuration.
                self._on_almost_full()
                continue

            # We're out of space. Raise an error.
            from hirola.exceptions import HashTableFullError
            source, value = self._blame_key(index, keys, shape)
            raise HashTableFullError(
                f"Failed to add {source} = {value} to the "
                f"hash table because the table is full and {value} "
                f"isn't already in it.")

    @property
    def almost_full(self):
        """The response to an almost full hash table. Hash tables become
        dramatically slower, the closer they get to being full. Hirola's default
        behaviour is to warn if this happens but can be configured to ignore the
        warning, raise an error or automatically make a new, larger table.

        This is an overloaded parameter.

        * :py:`almost_full = None`:
            Disable the *almost full* warning entirely.
        * :py:`almost_full = (0.8, "warn")`:
            Issue a `hirola.exceptions.AlmostFull` warning if the table reaches
            80% full.
        * :py:`almost_full = (0.7, "raise")`:
            Raise a `hirola.exceptions.AlmostFull` exception if the table
            reaches 80% full.
        * :py:`almost_full = (0.7, 2)`:
            Whenever the table reaches 70% full, double the table size.

        For reference, Python's `dict` grows 8-fold when two thirds full.
        To mimic this behaviour, set :py:`table.almost_full = (2 / 3, 8)`.

        """
        return self._almost_full

    @almost_full.setter
    def almost_full(self, x):
        # Asides from simply storing the user's value, this setter must also:
        # * Calculate the "panic table length" (self._raw.panic_at) at which the
        #   C code should notify Python that the table is almost full.
        # * Ensure that the user input is valid.

        if x is None:
            self._raw.panic_at = -1
            self._almost_full = None
            return

        try:
            ratio, scale_up = x
        except:
            raise TypeError(f"`almost_full` must be a 2-tuple of floats or"
                            f" None. Not `{repr(x)}`.") from None
        if not (0 < ratio <= 1):
            raise ValueError("The first parameter to almost_full must be "
                             ">0 and <=1.")
        if isinstance(scale_up, str):
            if scale_up not in ("raise", "warn"):
                raise ValueError("Valid near-full actions are 'raise' and "
                                 f"'warn'. Not '{scale_up}'.")
        elif isinstance(scale_up, numbers.Number):
            if int(self.max * scale_up) <= self.max:
                raise ValueError(
                    f"A scale_up resize factor of {scale_up} would lead to an "
                    f"infinite loop. Either increase scale_up or disable "
                    f"automatic resizing by setting hash_table.almost_full to "
                    f"None.")
        else:
            raise TypeError("The second parameter to almost_full must be "
                            "either a string or a float.")

        self._raw.panic_at = int(math.ceil(ratio * self.max))
        self._almost_full = x

    def _on_almost_full(self):
        """The callback to be invoked whenever the table becomes almost full."""
        assert self.almost_full is not None

        if not isinstance(self.almost_full[1], str):
            self.resize(self.max * self.almost_full[1], in_place=True)
            return

        from hirola.exceptions import AlmostFull

        message = f"HashTable() is {round(100 * len(self) / self.max)}% full." \
                  " A hash table becomes orders of magnitudes slower " \
                  "when nearly full. See help(HashTable.almost_full) for how " \
                  "to correct or silence this issue."

        if self.almost_full[1] == "raise":
            raise AlmostFull(message)
        import warnings
        warnings.warn(AlmostFull(message), stacklevel=3)

    def contains(self, keys) -> Union[bool, np.ndarray]:
        """Check if a key or keys are in the table.

        Args:
            keys:
                Elements to check for.
        Returns:
            Either true or false for each key in **keys**.

        This function is equivalent to but faster than
        :py:`table.get(keys) != -1`.
        To check only one key you may also use :py:`key in table`.

        """
        self._check_destroyed()

        keys, shape = self._norm_input_keys(keys)
        out = np.empty(shape, bool)
        slug.dll.HT_contains(self._raw._ptr, ptr(keys), ptr(out), out.size)
        return out if shape else out.item()

    __contains__ = contains

    def get(self, keys, default=-1) -> np.ndarray:
        """Lookup indices of **keys** in `keys`.

        Arguments:
            keys:
                Elements to search for.
            default:
                Returned inplace of a missing key.
                May be any object.
        Returns:
            The index/indices of **keys** in this table's `keys`. If a
            key is not there, returns :py:`-1` in its place.
        Raises:
            ValueError:
                If the :attr:`~numpy.ndarray.dtype` of **keys** doesn't match
                the `dtype` of this table.
            exceptions.HashTableDestroyed:
                If the `destroy` method has been previously called.

        """
        keys, shape = self._norm_input_keys(keys)
        out = np.empty(shape, np.intp)
        # This function forks out to several similar C functions depending on
        # how missing keys are to be handled.

        if default is self._NO_DEFAULT:
            # Default disabled - raise a key error if anything is missing.
            index = slug.dll.HT_gets_no_default(self._raw._ptr, ptr(keys),
                                                ptr(out), out.size)
            if index != -1:
                source, value = self._blame_key(index, keys, shape)
                raise KeyError(f"{source} = {value} is not in this table.")

        elif isinstance(default, numbers.Integral):
            if default == -1:
                # The default behaviour - use -1 to indicate missing keys.
                # This is already how the underlying C functions communicate
                # missing keys so nothing special needs to be done.
                slug.dll.HT_gets(self._raw._ptr, ptr(keys), ptr(out), out.size)
            else:
                # Not the default of -1 but still an integer default which can
                # be handled faster in C.
                slug.dll.HT_gets_default(self._raw._ptr, ptr(keys), ptr(out),
                                         out.size, default)

        else:
            # The slowest case: Return some non integer user defined default.
            slug.dll.HT_gets(self._raw._ptr, ptr(keys), ptr(out), out.size)
            out = np.where(out == -1, default, out)

        return out if shape else out.item()

    def _blame_key(self, index, keys, shape) -> Tuple[str, str]:
        """Get a key and its location from a ravelled index. Used to prettify
        key errors."""
        assert index >= 0
        if len(shape) == 0:
            if self._dtype_shape:
                return "key", repr(keys)
            return "key", repr(keys.item())
        if len(shape) == 1:
            return f"keys[{index}]", repr(keys[index])
        index = np.unravel_index(index, shape)
        return ("keys[" + ', '.join(map(str, index)) + "]"), repr(keys[index])

    def __getitem__(self, key):
        return self.get(key, default=self._NO_DEFAULT)

    def _check_dtype(self, keys):
        keys = np.asarray(keys, order="C")
        if keys.dtype != self._base_dtype:
            raise TypeError(
                "The dtype must match the dtype of the hash table. Expecting {}"
                " but got {}.".format(self._base_dtype, keys.dtype))
        return keys

    def _check_str_dtype(self, keys):
        return np.asarray(keys, dtype=self.dtype, order="C")

    def _norm_input_keys(self, keys):
        """Prepare input to be fed to C.

        * Convert to C contiguous array if not already.
        * Check/raise for wrong dtype.
        * For unlabelled records dtypes, right-strip self._dtype_shape.

        """
        self._check_destroyed()
        keys = self._check_dtype(keys)

        if self._dtype_shape:
            split = keys.ndim - len(self._dtype_shape)
            if keys.shape[split:] != self._dtype_shape:
                raise ValueError(
                    "The given shape's suffix doesn't match that of the hash "
                    "table's dtype. Received {} which doesn't end with {}.".
                    format(self._dtype_shape, keys.shape))
            return keys, keys.shape[:split]
        return keys, keys.shape

    def destroy(self) -> np.ndarray:
        """Release a writeable version of `keys` and permanently disable
        this table.

        Returns:
            A writeable shallow copy of `keys`.

        Modifying `keys` would cause an internal meltdown and is
        therefore blocked by setting the writeable flag to false. However, if
        you no longer need this table then it is safe to do as you please with
        the `keys` array. This function grants you full access to
        `keys` but blocks you from adding to or getting from this table
        in the future. If you want both a writeable `keys` array and
        functional use of this table then use :py:`table.keys.copy()`.

        """
        # Destruction is just setting a flag.
        self._destroyed = True
        return self._keys[:self.length]

    def _check_destroyed(self):
        if self._destroyed:
            from hirola.exceptions import HashTableDestroyed
            raise HashTableDestroyed

    def resize(self, new_size, in_place=False) -> 'HashTable':
        """Copy the contents of this table into a new `HashTable` of a
        different size.

        Args:
            new_size:
                The new value for `max`.
            in_place:
                If true resize this table. Otherwise make a modified copy.
        Returns:
            Either a new hash table with attributes `keys` and `length`
            matching those from this table or this table.
        Raises:
            ValueError:
                If requested size is too small (:py:`new_size < len(table)`).

        .. versionchanged:: 0.3.0
            Add the **in_place** option.

        """
        self._check_destroyed()
        if new_size < self.length:
            raise ValueError(f"Requested size {new_size} is too small to fit "
                             f"{self.length} keys.")
        new = type(self)(new_size, self.dtype, almost_full=self.almost_full)
        # This is only marginally (10-20%) faster than just creating an empty
        # table and running `new.add(self.keys)`.
        slug.dll.HT_copy_keys(self._raw._ptr, new._raw._ptr)

        if in_place:
            # In place resizing really just creates a resized copy then moves
            # all the resized attributes back to the original.
            self._raw = new._raw
            self._keys = new._keys
            self._hash_owners = new._hash_owners
            self._keys_readonly = new._keys_readonly
            self.almost_full = new.almost_full
            return self

        return new

    def copy(self, usable=True) -> 'HashTable':
        """Deep copy this table.

        Args:
            usable:
                If set to false and this table has called `destroy`, then
                the destroyed state is propagated to copies.
        Returns:
            Another `HashTable` with the same size, dtype and content.

        """
        out = type(self)(self.max, self.dtype, self.almost_full)
        if self._destroyed and usable:
            out.add(self.keys)
        else:
            out._raw.length = self._raw.length
            out._destroyed = self._destroyed
            out._hash_owners[:] = self._hash_owners
            out._keys[:out._raw.length] = self.keys

        return out


def choose_hash(key_size):
    if key_size < 4:
        hash = slug.dll.small_hash
    elif key_size % 4 == 0:
        hash = slug.dll.hash
    else:
        hash = slug.dll.hybrid_hash
    return hash


def vectorise_hash(hash, key_size, keys):
    """Apply a hash() function to an array of **keys**. Only used for testing.
    """
    keys = np.ascontiguousarray(keys)
    out = np.empty(keys.size * keys.dtype.itemsize // key_size, dtype=np.int32)
    slug.dll.vectorise_hash(ctypes.cast(hash, ctypes.c_void_p), ptr(keys),
                            ptr(out), key_size, out.size)
    return out
